intra/dialers/weird_split.go

```diff
@@ -0,0 +1,88 @@
+package dialers
+
+import (
+	"io"
+	"net"
+	"strings"
+	"syscall"
+)
+
+type WeirdSplitter struct {
+	*net.TCPConn
+	Used bool
+	Size int
+}
+
+func (s *WeirdSplitter) Write(b []byte) (int, error) {
+	conn := s.TCPConn
+	if s.Used {
+		// After the first write, there is no special write behavior.
+		return conn.Write(b)
+	}
+
+	// Setting `used` to true ensures that this code only runs once per socket.
+	s.Used = true
+	if len(b) < 2 {
+		return conn.Write(b)
+	}
+	var size int
+	if len(b) > s.Size {
+		size = s.Size
+	} else {
+		size = 1
+	}
+	rawConn, err := conn.SyscallConn()
+	if err != nil {
+		return 0, err
+	}
+	isIPv6 := strings.Contains(conn.RemoteAddr().String(), "[")
+
+	rawErr := rawConn.Control(func(fd uintptr) {
+		if isIPv6 {
+			err = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_IPV6, syscall.IPV6_UNICAST_HOPS, 1)
+		} else {
+			err = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_IP, syscall.IP_TTL, 1)
+		}
+	})
```

Comment:
Can you document a bit on what this approach does to bypass censorship *or* link to existing projects using this technique?

Is it similar to what's discussed here? #46

Reply:
[GoodbyeDPI commit e28cb526](https://github.com/ValdikSS/GoodbyeDPI/commit/e28cb52645f539dd044a21ce0fd0c015b4ffb91f) 
ValdikSS wrote:
>Some websites (or more precisely, TLS terminators/balancers) can't
handle segmented TLS ClientHello packet properly, requiring the whole
ClientHello in a single segment, otherwise the connection gets dropped.

>However they still operate with a proper TCP stack.
Cheat on them: send the latter segment first (with TCP SEQ "in the future"),
the former segment second (with "current" SEQ), allowing OS TCP
stack to combine it in a single TCP read().

But we don't have privileges, so we should implement this with setting TTL.  The TCP stack transmits the segment with TTL=1 first, then we restore TTL option before the TCP stack retransmits, this idea is mentioned in #46 too.